openshift/console
# Bug 2074767: change metrics queries based on metrics level configurations

**Fixes**: 

 https://bugzilla.redhat.com/show_bug.cgi?id=2074767


**Analysis / Root cause**: 

 Metrics page uses that pipelinerun/taskrun level metrics (histogram values) but with OSP 1.6.x and higher, the metrics level is set to pipeline/task and pipelinerun and taskrun histogram values are not supported.
 
 Ref: https://github.com/tektoncd/pipeline/blob/main/docs/metrics.md 

---

**Solution Description**: 

Choose different query set based on the metrics level configuration.

1. Show top 2 charts for the default pipeline/task metrics level and histogram type
2. Show all 4 charts for the pipelinerun/taskrun metrics level and lastvalue type
3. ~Hide metrics tab if the user don't have view access in openshift-pipelines (TBD)~ - Adding `honorLabels` to pipeline Service Monitor gets us the metric results and its fixed by operator team
4. Add Quickstart link to guide the users to change the metrics level configurations.

**Note: This solution works for both kubeadmin/non-kubeadmin users**

---
**Screen shots / Gifs for design review**: 


Default pipeline metrics level **pipeline/task** with QS link for kubeadmin/ns-admins:

![image (4)](https://user-images.githubusercontent.com/9964343/170243777-1f6fda31-2302-4bdf-b559-38a054fe1ff6.png)

Default pipeline metrics level view for consoledeveloper/non-admins (without QS):

![image (5)](https://user-images.githubusercontent.com/9964343/170244045-e8ed1067-1ea9-4843-8d43-38cb1ae89e08.png)

Pipeline Metrics level set to **pipelinerun/taskrun**:

![image](https://user-images.githubusercontent.com/9964343/170243463-6876ea72-744e-4f65-bbd8-49abf16129dc.png)

Quickstart to guide the user to change metrics level:
![image](https://user-images.githubusercontent.com/9964343/170244432-f726719a-3672-4ef0-b4fa-6d718441463f.png)

**Unit test coverage report**: 

```
  Pipeline Details page:

  Pipeline Details page Metrics Tab:
    ✓ It should contain metrics tab if the user has view permission in the openshift pipelines namespace (16ms)
    ✓ It should not contain metrics tab if the user does not have view access to openshift-pipelines namespace (14ms)
    ✓ It should contain metrics tab if the user has permission and osp 1.5.2 is installed (16ms)

    PipelineMetrics.spec.tsx:
      ✓ Should render only success ratio and number of pipeline runs charts if the metrics level is set to pipeline /task (default) level (13ms)
     ✓ Should contain quickstart link if the metrics level is set to pipeline /task (default) level (2ms)
     ✓ Should contain quickstart link if the user has update permission (2ms)
     ✓ Should not contain quickstart link if the user does not have update permission (1ms)
     ✓ Should not contain quickstart link if the metrics level is set to pipelinerun /taskrun level (2ms)


   pipeline-operator.spec.ts: 
    isSimplifiedVersionInstalled:
      ✓ should return false if the operator is not identified
      ✓ should return false if the installed operator is less than or equal to 1.5.2 (1ms)
      ✓ should return true if the installed operator is above 1.5.2

   pipeline-config.spec.ts:

     getPipelineMetricsLevel:
      ✓ should return unsupported level for the invalid configuration combinations (2ms)
      ✓ should return unsupported level if the config is not passed (2ms)
      ✓ should return unsupported level for null or invalid values
      ✓ should return pipeline/task level for the pipeline task default configuration
      ✓ should return pipelinerun/taskrun level for the pipelinerun/taskrun and lastvalue as duration type (1ms)
```


**Test setup:**


Prerequisite:

1. Add ConsoleQuickstart should be added to the cluster, use this [yaml](https://github.com/tektoncd/operator/blob/main/cmd/openshift/operator/kodata/tekton-addon/optional/quickstarts/configure-pipeline-metrics-quickstart-yaml) 
2. Add `honorLabels: true` in the pipeline service monitor like it is mentioned in this [operator PR](https://github.com/tektoncd/operator/pull/799) 
3. Add a workload from Git Import flow with Add pipeline options selected.
4. Visit the pipeline details page.


**Browser conformance**:
